File: examples/python/agent/shell.py
"""Demo agent: LangGraph agent with shell execution (demonstrates approval flow).

The shell.exec tool is governed by policy:
  - Safe commands (ls, cat, pwd, ...) are auto-allowed by the agent policy
  - Other commands require human approval via the client
  - The global policy denies shell.exec for any agent without explicit rules
"""

import asyncio
import logging
import os
from typing import Any

from langchain_openai import ChatOpenAI
from langchain.agents import create_agent

from rebuno import AsyncAgentContext, AsyncBaseAgent

logging.basicConfig(
    level=logging.INFO, format="%(asctime)s %(name)s %(levelname)s %(message)s"
)
logger = logging.getLogger("demo-agent-shell")

SYSTEM_PROMPT = (
    "You are a helpful sysadmin assistant. You can run shell commands to answer "
    "questions about the system. Use the shell tool to execute commands. Start with "
    "safe read-only commands when possible."
)


class ShellAgent(AsyncBaseAgent):
    """LangGraph agent that executes shell commands under kernel policy."""

    def __init__(self, **kwargs: Any):
        super().__init__(**kwargs)
        self._model = os.environ.get("OPENAI_MODEL", "gpt-4o-mini")

    async def process(self, ctx: AsyncAgentContext) -> dict:
        query = ""
        if isinstance(ctx.input, dict):
            query = ctx.input.get("query", "")
        elif isinstance(ctx.input, str):
            query = ctx.input

        if not query:
            return {"error": "No query provided"}

        logger.info("Processing query: %s", query)

        tools = ctx.get_tools()

        llm = ChatOpenAI(model=self._model, temperature=0)
        agent = create_agent(
            model=llm,
            tools=tools,
            system_prompt=SYSTEM_PROMPT,
        )

        result = await agent.ainvoke(
            {"messages": [{"role": "user", "content": query}]}
        )

        final_message = result["messages"][-1].content
        logger.info("Agent finished: %s", final_message)

        return {"query": query, "answer": final_message}


agent = ShellAgent(
    agent_id="shell-assistant",
    kernel_url=os.environ.get("REBUNO_KERNEL_URL", "http://localhost:8080"),
    api_key="your-secret-token",
)

@agent.tool("shell.exec")
async def shell_exec(command: str, timeout: int = 30) -> dict:
    """Execute a shell command."""
    logger.info("Executing: %s", command)
    try:
        proc = await asyncio.create_subprocess_shell(
            command,
            stdout=asyncio.subprocess.PIPE,
            stderr=asyncio.subprocess.PIPE,
        )
        stdout, stderr = await asyncio.wait_for(
            proc.communicate(), timeout=timeout
        )
        return {
            "exit_code": proc.returncode,
            "stdout": stdout.decode(errors="replace"),
            "stderr": stderr.decode(errors="replace"),
        }
    except asyncio.TimeoutError:
        proc.kill()
        return {"exit_code": -1, "stdout": "", "stderr": "Command timed out"}


if __name__ == "__main__":
    asyncio.run(agent.run())

File: examples/typescript/agent/shell.ts
// Demo agent: LangGraph agent with shell execution (demonstrates approval flow).
//
// The shell.exec tool is governed by policy:
//   - Safe commands (ls, cat, pwd, ...) are auto-allowed by the agent policy
//   - Other commands require human approval via the client
//   - The global policy denies shell.exec for any agent without explicit rules

import { exec } from "node:child_process";
import { z } from "zod";
import { ChatOpenAI } from "@langchain/openai";
import { createReactAgent } from "@langchain/langgraph/prebuilt";
import { BaseAgent, AgentContext, defineTool } from "rebuno";
import { toLangchainTools } from "rebuno/tools/adapters/langchain";

const SYSTEM_PROMPT =
  "You are a helpful sysadmin assistant. You can run shell commands to answer " +
  "questions about the system. Use the shell tool to execute commands. Start with " +
  "safe read-only commands when possible.";

// LangGraph agent that executes shell commands under kernel policy.
class ShellAgent extends BaseAgent {
  private model = process.env.OPENAI_MODEL ?? "gpt-4o-mini";

  async process(ctx: AgentContext) {
    const input = ctx.input as Record<string, unknown> | string;
    const query =
      typeof input === "object" && input !== null
        ? (input.query as string) ?? ""
        : String(input);

    if (!query) return { error: "No query provided" };

    console.log(`Processing query: ${query}`);

    const tools = await toLangchainTools(ctx.getTools()) as any;

    const llm = new ChatOpenAI({ model: this.model, temperature: 0 });
    const agent = createReactAgent({
      llm,
      tools,
      stateModifier: SYSTEM_PROMPT,
    });

    const result = await agent.invoke({
      messages: [{ role: "user", content: query }],
    });

    const finalMessage = result.messages[result.messages.length - 1].content;
    const answer =
      typeof finalMessage === "string"
        ? finalMessage
        : JSON.stringify(finalMessage);
    console.log(`Agent finished: ${answer.slice(0, 200)}`);

    return { query, answer };
  }
}

const agent = new ShellAgent({
  agentId: "shell-assistant",
  kernelUrl: process.env.REBUNO_KERNEL_URL ?? "http://localhost:8080",
});

agent.addTool(
  defineTool({
    id: "shell.exec",
    description: "Execute a shell command.",
    inputSchema: z.object({
      command: z.string(),
      timeout: z.number().default(30),
    }),
    execute: async ({ command, timeout = 30 }) => {
      console.log(`Executing: ${command}`);
      const timeoutMs = timeout * 1000;
      return new Promise<Record<string, unknown>>((resolve) => {
        const child = exec(
          command,
          { timeout: timeoutMs },
          (error: Error | null, stdout: string, stderr: string) => {
            if (
              error &&
              (error as NodeJS.ErrnoException & { killed?: boolean }).killed
            ) {
              resolve({
                exitCode: -1,
                stdout: "",
                stderr: "Command timed out",
              });
              return;
            }
            resolve({
              exitCode: (error as NodeJS.ErrnoException)?.code ?? 0,
              stdout: stdout ?? "",
              stderr: stderr ?? "",
            });
          },
        );
        setTimeout(() => child.kill(), timeoutMs);
      });
    },
  }),
);

agent.run();
